File: 07.Circular Queue/circular_queue.py
import array


class Queue:

    def __init__(self):
        self.q_size = 5
        self._data = array.array('b')
        self._tail = 0
        self._head = 0

    def get_tail(self):
        return self._tail

    def get_head(self):
        return self._head

    def enqueue(self, item):
        if (self._tail + 1) % (self.q_size + 1) == self._head:
            print("Queue is full!")
            return

        self._data.insert(self._tail, item)
        self._tail = (self._tail + 1) % (self.q_size + 1)

    def dequeue(self):
        if self._tail == self._head:
            print("Queue is empty")
            return -1

        item = self._data[self._head]
        self._head = (self._head + 1) % (self.q_size + 1)
        return item


if __name__ == "__main__":
    my_queue = Queue()
    
    my_queue.enqueue(1)
    print(f"Tail: {my_queue.get_tail()}")
    my_queue.enqueue(2)
    print(f"Tail: {my_queue.get_tail()}")

    print(f"Begining Head: {my_queue.get_head()}")
    item = my_queue.dequeue()
    print(f"Begining Head: {my_queue.get_head()}")
    item = my_queue.dequeue()
    print(f"Begining Head: {my_queue.get_head()}")
    item = my_queue.dequeue()

File: 07.Circular Queue/circular_queue.c
#include <stdio.h>
#define Q_SIZE 5

typedef struct {
    int data[Q_SIZE + 1];
    int head, tail;
} Queue;


void enqueue(Queue *q, int item);
int dequeue(Queue *q);


int main(){

    Queue my_queue; int item;

    my_queue.head = 0;
    my_queue.tail = 0;

    enqueue(&my_queue, 1);
    printf("tail = %d\n", my_queue.tail);
    enqueue(&my_queue, 2);
    printf("tail = %d\n", my_queue.tail);

    printf("beginning head = %d\n", my_queue.head);
    item = dequeue(&my_queue);

    printf("beginning head = %d\n", my_queue.head);
    item = dequeue(&my_queue);

    printf("beginning head = %d\n", my_queue.head);
    item = dequeue(&my_queue);

    return 0;
}


void enqueue(Queue *q, int item){
    
    if ((q->tail + 1) % (Q_SIZE + 1) == q->head)
    {
        printf("Queue if full!\n");
        return;
    }

    q->data[q->tail] = item;
    q->tail = (q->tail + 1) % (Q_SIZE + 1);
}


int dequeue(Queue *q){
    int item;
    if (q->tail == q->head)
    {
        printf("Queue is empty\n");
        return -1;
    }

    item = q->data[q->head];
    q->head = (q->head + 1) % (Q_SIZE + 1);

    return item;
}